Lint only the imported files

jsg2021 opened the pull request:
This is attempting to implement ideas in #23

This PR contains a:

- [ ] **bugfix**
- [x] new **feature**
- [x] **code refactor**
- [x] **test update** <!-- if bug or feature is checked, this should be too -->
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

Allow the linter to run on only files imported into the module graph. Ignoring any other files.

### Breaking Changes

None so far. However, this change will likely make the existing paradigm redundant.

### Additional Info

~~I'm just opening this to get started on it. It's not done.~~ Looking at the original implementation, this feature feels like a mental shift and may make sense to remove the existing "what to lint" logic to further simplify the configuration. idk, thoughts?

jsg2021 commented:
I went ahead and just refactored this to run on the import graph. So, not running on all files in the project/context could be considered a breaking change.

ricardogobbosouza commented:
Hi @jsg2021 
It looks incredible! This improves performance dramatically and I believe it solves #30

ricardogobbosouza commented:
/cc @evilebottnawi

evilebottnawi commented on src/linter.js:
```diff
@@ -28,14 +29,19 @@ export default async function linter(options, compiler) {
 
   try {
+    if (files.length === 0) {
+      throw new Error('No files matching');
```
Not sure it should be the error, when nothing to lint we should do nothing

jsg2021 replied:
I was just trying to make this pass existing tests. I agree, its odd. I'll remove it.

evilebottnawi replied:
Feel free to correct tests too

evilebottnawi commented on src/linter.js:
```diff
@@ -64,37 +70,42 @@ export default async function linter(options, compiler) {
 
-  compiler.hooks.emit.tapPromise('ESLintWebpackPlugin', async (compilation) => {
-    const { outputReport } = options;
+  compiler.hooks.afterEmit.tapPromise(
```
Ideally it should be: 

```
compiler.hooks.thisCompilation.tap((compilation) => { 
  const Compilation = compiler.webpack.Compilation;
  compilation.hooks.process.processAssets({ 
    name: 'name', 
    stage:  Compilation.PROCESS_ASSETS_STAGE_REPORT}, () => {
      // Logic
    }) 
})`
```

https://github.com/webpack/webpack/blob/master/lib/Compilation.js#L3470

for wepback@4 we can use https://github.com/webpack/webpack/blob/master/lib/Compilation.js#L514, 

It allows other plugins to use report file for other purposes (rare case, but it is good to support this)

jsg2021 replied:
is there an example I can reference?

jsg2021 replied:
I've attempted to split concerns some so it would be easier to switch methods of emitting the report. However, the webpack api is new to me. I'm trying to implement your suggestions, but I'm fairly certain I am doing it wrong.

jsg2021 replied:
@evilebottnawi `compilation.hooks.process` is undefined in webpack 5. `compilation.hooks.processAssets` is not a function. What am I missing? `compilation.hooks.processAssets.tap()`?  Stage doesn't seem to be honored?

evilebottnawi commented on src/index.js:
```diff
@@ -30,39 +30,44 @@ class ESLintWebpackPlugin {
       files: parseFiles(this.options.files, this.getContext(compiler)),
       extensions: arrify(this.options.extensions),
     };
+    Object.freeze(options);
```
Why?

jsg2021 replied:
ah, this was leftover from an exploration. I'll remove it.

evilebottnawi commented on src/index.js:
```diff
@@ -30,39 +30,44 @@ class ESLintWebpackPlugin {
 
-      /* istanbul ignore next */
-      compiler.hooks.watchRun.tapPromise(
+      runCompiler.hooks.compilation.tap(
```
`thisCompilation`, otherwise we will run plugin on child compilations, but we should do not do it

jsg2021 replied:
what should I do for webpack 4?

evilebottnawi commented on src/index.js:
```diff
@@ -30,39 +30,44 @@ class ESLintWebpackPlugin {
+    if (!options.lintDirtyModulesOnly) {
+      compiler.hooks.run.tapPromise('ESLintWebpackPlugin', run);
     }
+    compiler.hooks.watchRun.tapPromise('ESLintWebpackPlugin', run);
```
We can look at `compiler.watching` (webpack@5) to detect watching and improve DX

jsg2021 replied:
should I test for that existence or not support v4?

evilebottnawi replied:
Ideally we should test it because we support v4, but manually tests are enough here

jsg2021 replied:
I meant at runtime, detect it exists. I'm trying to find documentation on how to use `compiler.watching`. Any pointers?

evilebottnawi replied:
`compiler.watching` is not documented yes, but you can use it for getting `watching` or determinate what we are in watch

jsg2021 replied:
I'm not as familiar with this api. How would I change this line to use it?

jsg2021 replied:
Is there an example of this?

jsg2021 replied:
Are you suggesting we do not tap into watchRun unless we are watching?

jsg2021 replied:
like this?
```js
    if (compiler.watchMode ?? compiler.watching) {
      compiler.hooks.watchRun.tapPromise(ESLINT_PLUGIN, run);
    }
```

evilebottnawi commented on src/linter.js:
```diff
@@ -64,37 +70,42 @@ export default async function linter(options, compiler) {
     'ESLintWebpackPlugin',
     async (compilation) => {
       if (warnings.length > 0) {
+        // @ts-ignore
```
Need avoid it

jsg2021 replied:
the type for those lists is WebpackError, which appears to be a private class, is there a trick to making tsc happy?

evilebottnawi replied:
we need to added this to webpack types

evilebottnawi reviewed (changes requested):
Reviewed, need some improvements

jsg2021 commented:
@evilebottnawi I'm trying to understand. I think I've got it in a place we can easily wrap/move parts into different hooks.  I think I may understand what you meant by `compiler.watching`... see my comment on that thread... but the `processAssets` has me confused.  I've updated the code to use webpack to emit the report file asset. I'm trying to understand how to use the new webpack5 API and the v4 fallback apis you referenced. I'm just not finding any good examples. 

There is a timing issue here, I believe. 
1. The list of files to check need to be gathered.  (compilation.hooks.succeedModule)
2. Linter runs on files to produce results. (compiler.hooks.make)
3. Results are interpreted (compiler.hooks.emit)
4. Report conditionally emitted. (compiler.hooks.emit)

Thoughts?

evilebottnawi commented on package.json:
```diff
@@ -48,10 +48,10 @@
     "schema-utils": "^3.0.0",
-    "@types/eslint": "^7.2.0"
+    "@types/eslint": "^7.2.0",
+    "webpack-sources": "^2.0.0"
```
Do not use webpack-source directly, you can get them from `compiler.webpack`

jsg2021 replied:
`compiler.webpack` is undefined on webpack 4.

evilebottnawi commented on src/index.js:
```diff
@@ -31,38 +33,69 @@ class ESLintWebpackPlugin {
+          lint(files)
+        );
+
+        runCompiler.hooks.emit.tapPromise(ESLINT_PLUGIN, async () => {
```
Why `emit`? I wrote above that we should use

jsg2021 replied:
I'm learning as I go. I didn't know where your snippet went or how to use it. Would it go here?

jsg2021 replied:
`compilation.hooks.additionalChunkAssets` fires really early and its sync. I need an async event before it and after all modules are found. `compilation.hooks.afterSeal` is async, and allows me to get the async stuff completed... but it fires after `additionalChunkAssets`. :|

evilebottnawi replied:
hm, we should not write assets in `emit` hook, we can use `compier.outputFilesystem` in this case

jsg2021 replied:
done.

evilebottnawi commented on src/index.js:
```diff
@@ -1,28 +1,46 @@
+    // From my testing of compiler.watch() ... compiler.watching is always
+    // undefined (webpack 4 doesn't define it either) I'm leaving it out
+    // for now.
+    compiler.hooks.watchRun.tapPromise(ESLINT_PLUGIN, this.run);
```
/cc @ricardogobbosouza What do you think about improving it for auto detecting watch only on webpack@5?

ricardogobbosouza replied:
@evilebottnawi Sorry, but I didn't understand your question lol

evilebottnawi reviewed (commented):
Looks good, hope we don't break nothing

ricardogobbosouza commented:
I'm going to do some tests on webpack 4 and 5

evilebottnawi commented:
@ricardogobbosouza Feel free to ping me

ricardogobbosouza commented:
Hi @jsg2021 
Both in webpack4 and 5 the logs duplicate

![Captura de Tela 2020-10-28 às 10 06 04](https://user-images.githubusercontent.com/13064722/97439883-831a4580-1905-11eb-801f-0b53d0f008be.png)

jsg2021 commented:
@ricardogobbosouza huh, I didn't expect the that. I've added a test to reproduce it and fixed it. 

This was caused because I assumed each run call from `run`/`watchRun` woud have fresh `compiler.hooks.thisCompilation.taps`... just like `compilation.hooks.*`... so each run added the tap again.

jsg2021 commented:
Is there a tap option to remove after called? or a way to remove a tap? My current solution is _okay_, but idk, feels dirty.

evilebottnawi commented:
>Is there a tap option to remove after called? or a way to remove a tap? My current solution is okay, but idk, feels dirty.

yes should avoid it, it is reduce performance

jsg2021 commented:
> > Is there a tap option to remove after called? or a way to remove a tap? My current solution is okay, but idk, feels dirty.
> 
> yes should avoid it, it reduces performance

Is that true for once per-run? If I were to use it, I would only use it at the top so that each top-level run only had one use of this plugin. My current solution has the same effect. Its just a bit hacky in my opinion.

ricardogobbosouza commented:
@jsg2021  Apparently now works as expected =)

@evilebottnawi  I wonder if we have to have an option to linter all files and not just imported ones 🤔 
On smaller projects it can be useful, WDYT?

jsg2021 commented:
> @evilebottnawi  I wonder if we have to have an option to linter all files and not just imported ones 🤔 
> 
> On smaller projects it can be useful, WDYT?

In my opinion, that's beyond the scope of webpack. Files not touched by the import graph shouldn't touched. If the user wants to lint the entire project then the linter itself can do that as a separate task.

evilebottnawi commented:
@ricardogobbosouza 

> @evilebottnawi I wonder if we have to have an option to linter all files and not just imported ones thinking
On smaller projects it can be useful, WDYT?

Ideally yes, but it is edge case

evilebottnawi reviewed (approved):
Let's try

naulacambra commented:
Why is this merge blocked?
This new plugin really slowed my workflow, and this improvement is more than desired

ricardogobbosouza commented:
Hi @naulacambra
This is not blocked, as it is a significant change I am doing more tests, I must launch today
